content/posts/2023-12-optimising-spark-memory-usage.md

```diff
@@ -0,0 +1,278 @@
+problem came to our attention: the spark-evaluate service was consuming too much memory. Before I
+describe our change, let me explain what this service does.
+
+SPARK operates in rounds; each round takes approximately 1 hour in the current configuration. During
+the round, SPARK checker nodes running in Stations perform retrieval checks and submit the results -
+we call these records _measurements_. After a round is over, the evaluation service processes these
+measurements to evaluate the impact of each node (each participant). The Meridian smart contract
+uses the outcome of this evaluation to schedule rewards for node operators. You can learn more about
+Meridian on GitHub: https://github.com/Meridian-IE/impact-evaluator
+
+Here comes the issue: as the Station node network grows, so does the number of measurements
+submitted each round. Last week, we were evaluating more than 500 thousand measurements per round.
+Each measurement has around 600 bytes; we had to process 300MB of raw data every round.
+
+We implemented spark-evaluate to keep all data in memory because it was the fastest way to ship a
+working version. As we watched the network grow, every now and then we would receive an email from
+Fly.io that spark-evaluate crashed on an "out of memory" error. In the current stage, cloud
+resources are cheaper than engineering efforts, so we fixed these incidents by doubling the memory
+size of the virtual machine running the service. When we reached the point of 8 GB of RAM not being
+enough, we decided the situation was no longer sustainable, and the time came to do some engineering
+work.
+
+Fortunately, we found a small number of easy changes that significantly reduced our memory usage
+without any major changes in the implementation (like offloading the data from memory to the
+filesystem).
+
+## Plain data objects
+
+My colleague [Julian Gruber](https://github.com/juliangruber) discovered the first low-hanging
+fruit: apparently, plain data objects take more memory than class instances.
+
+Adding a step to map objects parsed from JSON to instances of a newly introduced `Measurement` class
+reduced our memory usage by 50%.
+
+The original version parsed measurements from JSON and then stored these objects for future
+evaluations.
+
+```js
+const measurements = /* fetch and parse JSON */
+const validMeasurements = measurements.map(measurement => {
+  try {
+    return {
+      ...measurement,
+      // additional fields parsed from data
+      }
+  } catch (err) {
+    logger.error('Invalid measurement:', err.message, measurement)
+    return null
+  }
+})
+```
+
+The new version converts each plain data object into a class instance. The class constructor
+performs preprocessing, e.g., parsing additional fields.
+
+```js
+const measurements = /* fetch and parse JSON */
+const validMeasurements = measurements.map(measurement => {
+  try {
+    return new Measurement(measurement)
```

Comment:
TODO: is it worth including the class definition of Measurement in the code snippet?